linkerd/stack/src/failfast.rs

```diff
@@ -43,112 +105,192 @@ pub enum ResponseFuture<F> {
-                    // The inner service just transitioned to NotReady, so initiate a new timeout.
-                    State::Open => {
+        let scope = self.scope;
+        loop {
+            match mem::replace(&mut self.state, State::Empty) {
+                State::Open(mut inner) => match inner.poll_ready(cx) {
+                    // The inner service just transitioned to `Pending`, so initiate a new timeout.
+                    Poll::Pending => {
                         self.wait
                             .as_mut()
                             .reset(Instant::now() + self.max_unavailable);
-                        debug!("{} service has become unavailable", self.scope);
-                        State::Waiting
+                        debug!("{scope} service has become unavailable");
+                        self.state = State::Waiting(inner)
                     }
-
-                    // A timeout has been set, so wait for it to complete.
-                    State::Waiting => {
+                    Poll::Ready(res) => {
+                        self.state = State::Open(inner);
+                        return Poll::Ready(res.map_err(Into::into));
+                    }
+                },
+                State::Waiting(mut inner) => match inner.poll_ready(cx) {
+                    // The inner service became ready, transition back to `Open`.
+                    Poll::Ready(res) => {
+                        trace!("{scope} has become ready");
+                        self.state = State::Open(inner);
+                        return Poll::Ready(res.map_err(Into::into));
+                    }
+                    // The inner service hasn't become ready yet --- has the
+                    // timeout elapsed?
+                    Poll::Pending => {
                         let poll = self.wait.as_mut().poll(cx);
+
                         if poll.is_pending() {
-                            trace!("Pending");
+                            // The timeout hasn't elapsed yet, keep waiting.
+                            trace!("{scope} service is Pending");
+                            self.state = State::Waiting(inner);
                             return Poll::Pending;
                         }
-                        warn!(
-                            "{} entering failfast after {:?}",
-                            self.scope, self.max_unavailable
-                        );
-                        State::FailFast
-                    }
 
-                    // Admit requests and fail them immediately.
-                    State::FailFast => {
+                        warn!("{scope} entering failfast after {:?}", self.max_unavailable);
+                        self.shared.in_failfast.store(true, Ordering::Release);
+
+                        let shared = self.shared.clone();
+                        self.state = State::FailFast(task::spawn(async move {
+                            let _ = inner.ready().await;
```

Comment:
Hmmm. Don't we need to propagate errors here?

Reply:
all this task is doing is notifying the `FailFast` service when the inner service becomes ready. when `FailFast` is notified, it will poll `inner` again. my assumption was that if poll_ready errors once, subsequent calls will also error.

we could change the task to propagate the error and not poll inner immediately after the `FailFast` is woken, but we would still have to return the inner service so that subsequent poll_readys can also get their own errors.

Reply:
okay, it returns the original error as of 990bf54219b31bed970e01addd02029ab7589acb

Reply:
990bf54219b31bed970e01addd02029ab7589acb changes this to eagerly propagate the original error, just to be safe.